### PR TITLE
Add support for http headers/parameters

### LINKED ISSUE #8: Add support for HTTP headers
Currently, the maven-download-plugin doesn't allow any HTTP header to be sent with query to server.
It would be really useful for cases where api secrets are required by servers (as a use case to download an example user JSON file from https://developer.wordpress.com/docs/api/1/get/me/)

### PR DESCRIPTION
**Pull request by Riduidel**
This pull request fixes #8 by replacing maven wagon by http client.
Unfortunatly, this impacts both `WGet` task and `ConsoleDownloadMonitor` which now uses information from http client stream.
This way, it is possible to inject during query construction the good headers and parameters.

### CODE REVIEW

**Comment by Riduidel** [comment]
I've tried to rewrite the git history for minimal impact. Unfortunatlyh, as I'm not a git wizard, I did it the hard way : deleted branch, and recreated it from the new source.

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java [hunk 1]
```diff
@@ -177,37 +191,55 @@
...
-        
+
         if (settings == null) {
             getLog().warn("settings is null");
         }
         getLog().debug("Got settings");
-        if (retries < 1) {
-            throw new MojoFailureException("retries must be at least 1");
-        }
+		if (retries < 1) {
```
Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: Would be nice if you could avoid such whitespace changes as they don't bring anything and create noise (and even inconsistency with the current usage there). It would be entirely fine to replace spaces by tabs, but it would be better in another commit.
(note that this is not a blocker for this patch to be merged; as the feature is valuable we wouldn't make a whitespace change blocking it).

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java [hunk 2]
```diff
@@ -323,50 +356,95 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
...
+	 * @param parameters http parameters used for that request
+	 * @param headers http headers used for that request
+	 * @param retries number of times that query will be retried
+	 * @throws Exception
+	 */
+	private void doGet(final URL source, final File outputFile, final Map<String, String> parameters, final Map<String, String> headers, final int retries) throws Exception {
         ConsoleDownloadMonitor downloadMonitor = null;
-        if (this.session.getSettings().isInteractiveMode()) {
-            downloadMonitor = new ConsoleDownloadMonitor(this.getLog());
-            wagon.addTransferListener(downloadMonitor);
```
Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: Should't this download monitor remain?

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java [hunk 3]
```diff
@@ -323,50 +356,95 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
...
-    private void doGet(File outputFile) throws Exception {
-        String urlStr = this.uri.toString();
-        String[] segments = urlStr.split("/");
-        String file = segments[segments.length - 1];
-        String repoUrl = urlStr.substring(0, urlStr.length() - file.length() - 1);
-        Repository repository = new Repository(repoUrl, repoUrl);
-
-        Wagon wagon = this.wagonManager.getWagon(repository.getProtocol());
-        if (readTimeOut > 0) {
-            wagon.setReadTimeout(readTimeOut);
```
Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: We need to keep the timeout feature.

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java [hunk 4]
```diff
@@ -323,50 +356,95 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
...
         ConsoleDownloadMonitor downloadMonitor = null;
-        if (this.session.getSettings().isInteractiveMode()) {
-            downloadMonitor = new ConsoleDownloadMonitor(this.getLog());
-            wagon.addTransferListener(downloadMonitor);
+        if (session.getSettings().isInteractiveMode()) {
+            downloadMonitor = new ConsoleDownloadMonitor(getLog());
         }
-
-        AuthenticationInfo authenticationInfo = new AuthenticationInfo();
-        if (StringUtils.isNotBlank(username)) {
```
Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: We need to keep the authentication feature.

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java [hunk 5]
```diff
@@ -323,50 +356,95 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
...
-            getLog().debug("providing custom authentication for " + serverId);
-            Server server = settings.getServer(serverId);
-            if (server == null)
-                throw new MojoExecutionException(String.format("Server %s not found", serverId));
-            getLog().debug(String.format("serverId %s supplies username: %s and password: ***",  serverId, server.getUsername() ));
-            authenticationInfo.setUserName(server.getUsername());
-            authenticationInfo.setPassword(server.getPassword());
-        }
-
-        ProxyInfo proxyInfo = this.wagonManager.getProxy(repository.getProtocol());
```
Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: We need to keep the proxy feature.

**Comment by Riduidel** [comment]
Sorry for the inconvenience.
I think I'll have to put all that code on the table and understand really what's going on.

**Comment by mickaelistria** [comment]
No inconvenience, don't worry, this is the usual lifecycle of a review process ;)
About the code, I have the impression that you should split it in 2 parts/PRs:
1. Update of httpclient and related code (Without adding anything else)
2. Addition of the `parameters` and `headers` with the related block of code. There have been no major architectural change between the older version and now, so I believe porting the feature is only about porting line 375..381 to the right place in the newer code, and updating the signature and consumers of `doGet` as you did.

**Comment by Riduidel** [comment]
Makes sense, I will do it as soon as possible.

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java (hunk 2 above)
Comment (Riduidel) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: Well, I choose to initialize the download monitor the same way it's currently done. If you look in current version of WGet#doGet, it's exactly what is done.

Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: Ok, I didn't check the current state of the code while reviewing the diff. Feel free to ignore this comment then.

File: download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java (hunk 4 above)
Comment (Riduidel) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: What kind of authentication is usable here ? preemptive basic ? or something more sophisticated ? http client supports many kinds of auth, but info is not clear on wagon side ... I "guess" it's preemptive basic, but can you confirm that guess ?

Comment (mickaelistria) [inline download-maven-plugin/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java]: Yes, I think so. At the moment, don't change that; but if you know better way to handle the multiple authentication strategies, that'd be welcome as a separate PR.

**Comment by mickaelistria** [comment]
FYI, instead of closing PRs and creating new ones, you can try the
following:
* Append a commit to your branch; it will be added to the PR
* Amend your commit locally by using `git commit --amend`, and then
overwrite your branch with `git push --force HEAD:rewrite-using-httpclient`.
That also updates the PR.
